frontends/p4/typeChecking/typeChecker.cpp

```diff
@@ -2448,19 +2472,51 @@ const IR::Node* TypeInference::postorder(IR::Member* expression) {
                 return expression;
             }
         }
-        if (type->is<IR::Type_Header>()) {
-            if (member == IR::Type_Header::setValid ||
-                member == IR::Type_Header::setInvalid) {
+        if (type->is<IR::Type_Header>() || type->is<IR::Type_HeaderUnion>()) {
+            if (type->is<IR::Type_HeaderUnion>() &&
+                (member == IR::Type_Header::setValid ||
+                member == IR::Type_Header::setInvalid)) {
+                ::error(ErrorType::ERR_INVALID, "setValid and setInvalid are "
+                        "not defined for header unions", expression);
+                return expression;
+            }
+            // Built-in methods
+            if ((member == IR::Type_Header::setValid ||
+                member == IR::Type_Header::setInvalid) &&
+                type->is<IR::Type_Header>()) {
                 if (!isLeftValue(expression->expr))
                     typeError("%1%: must be applied to a left-value", expression);
-                // Built-in method
                 auto type = new IR::Type_Method(IR::Type_Void::get(), new IR::ParameterList);
                 auto ctype = canonicalize(type);
                 if (ctype == nullptr)
                     return expression;
                 setType(getOriginal(), ctype);
                 setType(expression, ctype);
                 return expression;
+            } else if ((member == IR::Type_Header::sizeBits) ||
+                       (member == IR::Type_Header::sizeBytes)) {
+                int sz = 0;
+                auto t = typeMap->getType(expression->expr, true);
+                const IR::Type_StructLike* ht;
+                if (t->is<IR::Type_Header>()) {
+                    ht = t->to<IR::Type_Header>();
+                } else if (t->is<IR::Type_HeaderUnion>()) {
+                    ht = t->to<IR::Type_HeaderUnion>();
+                }
+                if (ht == nullptr) {
+                    ::error("%1%: null header/headerUnion?", expression);
+                    return expression;
+                }
+                sz = ht->width_bits();
```

Comment:
this won't work for Type_HeaderUnion. I think that the function width_bits() for Type_HeaderUnion is in fact buggy, since it calls width_bits on types which are all Type_Name.

Reply:
I don't see a Type_Name.  Please see this code.

https://github.com/p4lang/p4c/blob/master/ir/type.def#L258